internal/pkg/scorecard/scorecard.go

```diff
@@ -50,42 +54,99 @@ var (
+	if err := scViper.UnmarshalKey("plugins", &configs, func(c *mapstructure.DecoderConfig) { c.ErrorUnused = true }); err != nil {
+		return nil, errors.Wrap(err, "Could not load plugin configurations")
 	}
-	if viper.GetBool(scplugins.OLMTestsOpt) {
-		plugins = append(plugins, olmTestsPlugin)
+	for _, plugin := range configs {
+		if err := validateConfig(plugin); err != nil {
+			return nil, fmt.Errorf("error validating plugin config: %v", err)
+		}
+		var newPlugin Plugin
+		if plugin.Basic != nil {
+			pluginConfig := plugin.Basic
+			setConfigDefaults(pluginConfig, kubeconfig)
+			newPlugin = basicOrOLMPlugin{name: plugin.Name, pluginType: scplugins.BasicOperator, config: *pluginConfig}
+		} else if plugin.Olm != nil {
+			pluginConfig := plugin.Olm
+			setConfigDefaults(pluginConfig, kubeconfig)
+			newPlugin = basicOrOLMPlugin{name: plugin.Name, pluginType: scplugins.OLMIntegration, config: *pluginConfig}
+		} else {
+			setPaths = append(setPaths, plugin.External.Command)
+			pluginConfig := plugin.External
+			if kubeconfig != "" {
+				// put the kubeconfig flag first in case user is overriding it with an env var in config file
+				pluginConfig.Env = append([]externalPluginEnv{{Name: "KUBECONFIG", Value: kubeconfig}}, pluginConfig.Env...)
+			}
+			newPlugin = externalPlugin{name: plugin.Name, config: *pluginConfig}
+		}
+		// keep this statement after the previous if statements; otherwise default tests won't be disabled correctly
+		if plugin.Disable {
+			continue
+		}
+		plugins = append(plugins, newPlugin)
 	}
 	// find external plugins
-	pluginDir := viper.GetString(PluginDirOpt)
+	pluginDir := scViper.GetString(PluginDirOpt)
 	if dir, err := os.Stat(pluginDir); err != nil || !dir.IsDir() {
+		if setPaths != nil {
+			return nil, fmt.Errorf("plugin directory `%s` not found: %v", pluginDir, err)
+		}
 		log.Warnf("Plugin directory not found; skipping external plugins: %v", err)
-		return plugins
+		return plugins, nil
 	}
 	if err := os.Chdir(pluginDir); err != nil {
+		if setPaths != nil {
+			return nil, fmt.Errorf("could not change directory to plugin_dir `%s`: %v", pluginDir, err)
+		}
 		log.Warnf("Failed to chdir into scorecard plugin directory: %v", err)
-		return plugins
+		return plugins, nil
 	}
 	files, err := ioutil.ReadDir("bin")
 	if err != nil {
-		log.Errorf("Failed to list files in %s/bin; skipping external plugin tests: %v", pluginDir, err)
-		return plugins
+		log.Errorf("Failed to list files in %s/bin; skipping automatic external plugin tests: %v", pluginDir, err)
+		return plugins, nil
 	}
 	for _, f := range files {
-		plugins = append(plugins, genericPlugin{filepath.Join("./bin", f.Name())})
+		skip := false
+		for _, path := range setPaths {
+			if filepath.Join("bin", f.Name()) == path {
+				skip = true
+			}
+		}
+		if skip {
+			continue
+		}
+		plugins = append(plugins, externalPlugin{name: f.Name(), config: externalPluginConfig{Command: filepath.Join("bin", f.Name())}})
```

Comment:
Yes, that is correct. The reason we error out on point 2 is that we run all external plugins from the scorecard plugin directory. It is set to error because we can't run the plugins from where we say it should.

I agree with the idea that all external plugins should be explicitly defined in the config file now that we require a config file to run the scorecard. This then leads to another question. Should we still have a `plugin-dir` config option? We included it so we could automatically enumerate all existing external plugins. If we don't do that, this is probably unnecesssary and we can just use the project_dir as the default working directory for the plugins (or maybe we could add another config option for external plugins to allow configuration of the wd; maybe have a global wd for external plugins with the ability to override it on a plugin-by-plugin basis, similar to how `kubeconfig` and setting the kubeconfig env var for specific plugins is handled).

Reply:
> Should we still have a plugin-dir config option? We included it so we could automatically enumerate all existing external plugins. If we don't do that, this is probably unnecesssary and we can just use the project_dir as the default working directory for the plugins

Agreed. I'd say remove the `plugin-dir` option and use `project_dir` as the CWD. I think that would be more intuitive for plugin execution anyway. If plugins need a different working directory they can always `cd` internally.

Would it be easier to do that here or in a follow-up?

Reply:
It would be better to do it here

Reply:
Updated. The `plugin-dir` global opt is removed along with the `disable` plugin opt, which was only needed to disable auto-discovered plugins in `plugin-dir`